Implementation of MK docs.

RubelMozumder opened the pull request.

RubelMozumder committed:
add docs dir, rename READEME to index.

RubelMozumder committed:
Inclusing Structure of the mk docs.

RubelMozumder committed:
update pyproject.

RubelMozumder committed:
removing other empty md files from docs.

RubelMozumder committed:
commenting the other pages from mkdocs.yml

RubelMozumder committed:
First step for mydocs.

RubelMozumder committed:
Docs v1.

RubelMozumder committed:
v2

RubelMozumder committed:
Including mkdocs.yml

lukaspie commented on pyproject.toml:
```diff
@@ -43,6 +43,12 @@ dev = [
     "lxml-stubs",
+    "mkdocs",
+    # Mkdocs and the related plugins
+    "mkdocs-mermaid2-plugin"
```
```suggestion
    "mkdocs-mermaid2-plugin",
```

lukaspie commented on docs/explanations.md:
```diff
@@ -0,0 +1,223 @@
+# Motivation
+
+The NeXus data format, described by the NeXus Definition Language (NXDL), represents a concerted effort aimed at facilitating data exchange within scientific communities, particularly among those engaged in neutron, X-ray, and muon research [J. Appl. Cryst. (2015). 48, 301-305](https://doi.org/10.1107/S1600576714027575). The data format is also being used by the material science community under t ...
```
```suggestion
The NeXus data format, described by the NeXus Definition Language (NXDL), represents a concerted effort aimed at facilitating data exchange within scientific communities, particularly among those engaged in neutron, X-ray, and muon research [J. Appl. Cryst. (2015). 48, 301-305](https://doi.org/10.1107/S1600576714027575). The data format is also being used by the material science community under the project [NeXus-FAIRmat](https://fairmat-nfdi.github.io/nexus_definitions/) supporting FAIR (Findable, Accessible, Interoperable and Reuseable) data principles. It serves as a standardized framework for both data exchange and storage. At its core, the NeXus Definition Language (NXDL) functions as the cornerstone through which scientists delineate the nomenclature and organizational structure of information within NeXus data files, tailored to specific scientific techniques.
```

lukaspie commented on docs/explanations.md:
```diff
@@ -0,0 +1,223 @@
+
+To expedite the schema development process, we have introduced the use of Yet Another Markup Language ([YAML](https://yaml.org/)), which provides a syntax or style specifically tailored for defining scientific domain-driven schemas with NXDL. One significant advantage of YAML over XML is its indentation-driven approach, which eliminates the need for starting and ending tags for each entity within ...
+
+The `YAML` format, while not yet an official version of NeXus application definitions or base classes, necessitates a method for transcoding it into `XML`. The [nyaml](https://github.com/FAIRmat-NFDI/nyaml/tree/main) Python package serves as a converter tool designed specifically for this purpose. It enables the conversion of NXDL from `YAML` format to `XML`, thereby enhancing the capability of N ...
```
```suggestion
The `YAML` format, while not yet an official version of NeXus application definitions or base classes, necessitates a method for transcoding it into `XML`. The [nyaml](https://github.com/FAIRmat-NFDI/nyaml/tree/main) Python package serves as a converter tool designed specifically for this purpose. It enables the conversion of NXDL from `YAML` format to `XML`, thereby enhancing the capability of NeXus schema developers to incorporate domain-specific scientific knowledge into the schema. Furthermore, the tool offers the flexibility to extend existing NeXus schemas in XML by facilitating conversion back and forth between the two formats. It is important to note that here we do not introduce NeXus data objects, terms, or types, which are fundamental for writing base class schemas or application definition schemas. For individuals new to NeXus, we refer to the [official NeXus site](https://www.nexusformat.org/).
```

lukaspie commented on docs/index.md:
```diff
@@ -0,0 +1,45 @@
+
+Tutorials to write different parts and a full NeXus application or base class
+
+- [Tutorials to write NeXus definition in YAML](tutorials.md)
```
```suggestion
- [Tutorials for writing NeXus definition in YAML](tutorials.md)
```

lukaspie commented on docs/index.md:
```diff
@@ -0,0 +1,45 @@
+An introduction to NeXus and its design principles.
+
+- [An introduction to NeXus](https://manual.nexusformat.org/index.html)
+- [Motivation for NYAML tool](explanations.md)
```
```suggestion
- [Explanation of the nyaml tool](explanations.md)
```

lukaspie commented on docs/index.md:
```diff
@@ -0,0 +1,45 @@
+
+</div>
+<div markdown="block">
+### Reference
+Relavent references that supported to develop this tool:
+- [NYAML GitHub repository](https://github.com/FAIRmat-NFDI/nyaml)
+- [FAIRmat-NeXus website](https://fairmat-experimental.github.io/nexus-fairmat-proposal/)
+- [FAIRmat-NeXus GitHub repository](https://github.com/FAIRmat-NFDI/nexus_definitions)
+- [Official werbsite for NeXus](https://manual.nexusformat.org/index.html)
+
+- [NYAML developed under project FAIRmat-NFDI](https://github.com/FAIRmat-NFDI)
+- [National Research Data Infrastructure (NFDI) funded FAIRmat-NFDI](https://www.nfdi.de/?lang=en)
+</div>
+</div>
```
```suggestion
### References

- [NYAML GitHub repository](https://github.com/FAIRmat-NFDI/nyaml)
- [FAIRmat-NeXus website](https://fairmat-experimental.github.io/nexus-fairmat-proposal/)
- [FAIRmat-NeXus GitHub repository](https://github.com/FAIRmat-NFDI/nexus_definitions)
- [Official NeXus website](https://manual.nexusformat.org/index.html)
</div>
</div>

<h2>Project and community</h2>

Any questions or suggestions? [Get in touch!](https://www.fair-di.eu/fairmat/about-fairmat/team-fairmat)

[The work is funded by the Deutsche Forschungsgemeinschaft (DFG, German Research Foundation) - 460197019 (FAIRmat).](https://gepris.dfg.de/gepris/projekt/460197019?language=en)
```

lukaspie commented on tests/data/dimensions_parsed.yaml:
```diff
@@ -0,0 +1,25 @@
+category: application
```
These changes seem unconnected from the rest of the PR, is this a rebase issue?

lukaspie commented on docs/explanations.md:
```diff
@@ -0,0 +1,223 @@
+# Motivation
+
+The NeXus data format, described by the NeXus Definition Language (NXDL), represents a concerted effort aimed at facilitating data exchange within scientific communities, particularly among those engaged in neutron, X-ray, and muon research [J. Appl. Cryst. (2015). 48, 301-305](https://doi.org/10.1107/S1600576714027575). The data format is also being used by the material science community under t ...
```
I think most of this should actually go to the `index.md` file as an introduction to the tool.

lukaspie replied:
Basically, everything in this paragraph should just go to the index file.

lukaspie commented on docs/how_to_guide.md:
```diff
@@ -0,0 +1,39 @@
+# nyaml Installation and Command
+The `nyaml` is pathon package [published in PyPI](https://pypi.org/project/nyaml/) therefore easy to use and install.
```
```suggestion
`nyaml` is a Python package [published on PyPI](https://pypi.org/project/nyaml/).
```

lukaspie commented on docs/how_to_guide.md:
````diff
@@ -0,0 +1,39 @@
+```bash
+$ pip install nyaml
+```
+To contribute to the tool or to install it in development mode
````
```suggestion
To contribute to the tool or to install it in development mode, you should run
```

lukaspie commented on docs/how_to_guide.md:
````diff
@@ -0,0 +1,39 @@
+```
+
+## How to Use the nyaml Tool
+The `nyaml` works as a command line tool to convert NeXus application definition or base class from `yaml` file format into the `nxdl.xml` file format and vice-versa. The converter can be called by the command
````
```suggestion
`nyaml` works as a command line tool to convert NeXus application definitions or base classes from `yaml` file format into the `nxdl.xml` file format and vice-versa. The converter can be called by the command
```

lukaspie commented on pyproject.toml:
```diff
@@ -43,6 +43,12 @@ dev = [
+    "mkdocs-mermaid2-plugin"
+    "mkdocstrings",
+    "mkdocs-material"
+
```
add mkdocs-click, see comment in docs/how_to_guide.md

lukaspie commented on mkdocs.yml:
```diff
@@ -0,0 +1,44 @@
+  - pymdownx.snippets
+  - admonition
+  - pymdownx.details
+  - md_in_html
```
add mkdocs-click

lukaspie commented on docs/how_to_guide.md:
````diff
@@ -0,0 +1,39 @@
+## How to Use the nyaml Tool
+The `nyaml` works as a command line tool to convert NeXus application definition or base class from `yaml` file format into the `nxdl.xml` file format and vice-versa. The converter can be called by the command
+
+```bash
````
Use [`mkdocs-click`](https://github.com/mkdocs/) in the pyproject.toml file and then here, we can do
```
::: mkdocs-click
    :module: nyaml.cli
    :command: launch_tool
    :prog_name: nyaml2nxdl
    :depth: 2
    :style: table
    :list_subcommands: True
```

RubelMozumder replied:
Nice feature. Included.

lukaspie commented on mkdocs.yml:
```diff
@@ -0,0 +1,44 @@
+  - Home: index.md
+  - Tutorial: tutorials.md
+  - How-To-Guide : how_to_guide.md
+  - Reference: references.md
```
this file does not exist

lukaspie commented on docs/how_to_guide.md:
````diff
@@ -0,0 +1,39 @@
+                       assist in identifying issues in YAML files.
+  --help               Show this message and exit.
+```
+The `--output-file` option can be used to define the output file name (including the fle extension), otherwise the converter will define the output file name from the input file, e.g., for the input file `NXapplication.nxdl.xml (NXapplication.yaml)`, the resultant file will be `NXapplication_parser.yaml (NXapplication.nxdl.xml)`. With the option `--check-consistency` the converter produces the sa ...
````
```suggestion

- The `--output-file` option can be used to define the output file name (including the file extension), otherwise the converter will define the output file name from the input file, e.g., for the input file `NXapplication.nxdl.xml (NXapplication.yaml)`, the resultant file will be `NXapplication_parsed.yaml (NXapplication.nxdl.xml)`. 
- With the option `--check-consistency` the converter produces the same type of file as the input, e.g. for input `NXapplication.nxdl.xml`, the output file is `NXapplication_consistency.nxd.xml`. The intention for this option is to verify proper file and version conversion of the file. 
- When converting the `nxdl.xml` file into `yaml` it also stores the `nxdl.xml` file at the end of `yaml` file with a hash. The option `--do-not-store-nxdl` prevents the `yaml` file from storing the original `nxdl.xml` text. 
- The `verbose` option is to identify any issues arising from unexpected conversion or syntax errors that occur while converting the file from one to another.
```

What do you mean by "proper file and version conversion"?

lukaspie commented on docs/how_to_guide.md:
````diff
@@ -0,0 +1,39 @@
+  --help               Show this message and exit.
+```
+The `--output-file` option can be used to define the output file name (including the fle extension), otherwise the converter will define the output file name from the input file, e.g., for the input file `NXapplication.nxdl.xml (NXapplication.yaml)`, the resultant file will be `NXapplication_parser.yaml (NXapplication.nxdl.xml)`. With the option `--check-consistency` the converter produces the sa ...
+The `--output-file` option if user wants to define output file name (including extension) otherwise converter will define the output file name e.g. from input file `NXapplication.nxdl.xml (NXapplication.yaml)` the resultant file will be `NXapplication_parser.yaml (NXapplication.nxdl.xml)`. With the option `--check-consistency` the converter produces the same type of file as the input, e.g. for in ...
````
this seems to be duplicated, remove

lukaspie commented on docs/tutorials.md:
```diff
@@ -0,0 +1,185 @@
+# Tutorials
+This tutorial will explain different keywords, terms and rules from the prespective of YAML format of the NeXus schema.
```
```suggestion
This tutorial will explain different keywords, terms, and rules from the perspective of YAML format of the NeXus schema.
```

Honestly, this is not a tutorial. It should rather go to explanations.

lukaspie replied:
Note that tutorial, how-to guide, reference, etc. have special meaning in the [Diataxis framework](https://diataxis.fr/) we are using. This also means that the different sections should align with that meaning. But it's probably fine to keep it as is.

lukaspie commented on docs/tutorials.md:
```diff
@@ -0,0 +1,185 @@
+
+Descriptive information about NeXus groups is encapsulated within the `doc` child of the respective group. It is important to note that the group annotation of `source_TYPE(NXsource)` or `(NXsource)source_TYPE` signifies the encapsulation of the group's `name` as `source_TYPE` and its type as `NXsource` base class. Notably, the order between `name` and `type` within the XML element must be invert ...
+
+Furthermore, the uppercase part of the group's name can be dynamically overwritten, allowing for the instantiation of multiple instances. For example, `source_electric` and `source_magnetic` can coexist from `NXsource`. It is essential to adhere to the uppercase dynamic rules for NeXus groups, fields, and attributes.
```
This is now covered by `nameType=partial`, the documentation here should be updated.

lukaspie commented on docs/tutorials.md:
```diff
@@ -0,0 +1,185 @@
+### NeXus Choice
+NeXus `choice` concept is designed to choose a concept from a number of concepts of the same kind (e.g., a NeXus field). The `choice` options allows for defining a scientific concept in several modes for different situations (e.g., for different instrument configurations or measurement modes).
+
+**NeXus choice in YAML format**
```
This doesn't work yet, see https://github.com/FAIRmat-NFDI/nyaml/issues/23.

So, I don't think we should mention it here yet.

RubelMozumder replied:
Commented out!

lukaspie commented on docs/tutorials.md:
````diff
@@ -0,0 +1,185 @@
+  \@version:
+  enumeration: [NXmpes]
+```
+In the example the valid value for NeXus field `definition` is `NXmpes`.
````
```suggestion
In the example, the only valid value for NeXus field `definition` is `NXmpes`.
```

lukaspie commented on docs/tutorials.md:
```diff
@@ -0,0 +1,185 @@
+In the example the valid value for NeXus field `definition` is `NXmpes`.
+
+### Keyword `xref`
+The `xref` keyword (which can only inside the keyword `doc`) is used to refer any other ontology or any other standard such `ISO`. The `xref` in the example `doc` will reflect the information inside the XML `doc`. Note that the `xref` keyword is only available in the `YAML` representation and will be transformed into its textual representation inside the `doc` text in `XML`.
```
```suggestion
The `xref` keyword (which can only be used inside the keyword `doc`) is used to refer any other ontology or any other standard such `ISO`. The `xref` in the example `doc` will reflect the information inside the XML `doc`. Note that the `xref` keyword is only available in the `YAML` representation and will be transformed into its textual representation inside the `doc` text in `XML`.
```

lukaspie reviewed (changes requested):
Initial round of review, there are many places that could be improved. Would also be nice to improve the style a bit to be closer to the style of the documentation of our other tools.

lukaspie commented:
@RubelMozumder I think this PR is an excellent starting point for the paper and many of the materials here could be brought into the paper. For JOSS, it is also important to have good documentation, so getting this PR finished should be high on our priority list before submission.

RubelMozumder committed:
rebase1

RubelMozumder committed:
Commit rebase.

RubelMozumder committed:
Revision and inclustion of

RubelMozumder committed:
Update docs/index.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

Update docs/index.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

Update docs/index.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

Update docs/tutorials.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

PR change request.

Update docs/tutorials.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

Update mkdocs.yml

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

Update docs/how_to_guide.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

Update docs/how_to_guide.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

Update docs/how_to_guide.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

RubelMozumder committed:
Use mkdocs-click

RubelMozumder committed:
Being the README.md file.

RubelMozumder committed:
Remove old files.

RubelMozumder committed:
lint

Update docs/index.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

Update docs/index.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

Update mkdocs.yml

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

Update docs/tutorials.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

Update docs/tutorials.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

lukaspie commented on docs/index.md:
```diff
@@ -0,0 +1,53 @@
+
+How-to install and use the nyaml tool
+
+- [How-to guide for nyaml tools](how_to_guide.md)
```
```suggestion
- [How-to guide for using the `nyaml` tools(how_to_guide.md)
```

lukaspie commented on docs/index.md:
```diff
@@ -0,0 +1,53 @@
+
+Tutorials to write different parts and a full NeXus application or base class
+
+- [Tutorials for writing NeXus definition in YAML](tutorials.md)
```
```suggestion
- [Writing a NeXus definition in YAML](tutorials.md)
```

lukaspie commented on mkdocs.yml:
```diff
@@ -0,0 +1,45 @@
+  - Tutorial: tutorials.md
+  - How-To-Guide : how_to_guide.md
+  - Reference: references.md
+  - Explanation: explanations.md
```
```suggestion
  - Explanation: explanations.md
  - Reference: references.md
```
swapping the order to have them in the same way as they are showing up on the main page

lukaspie commented on .github/workflows/build_docs.yml:
```diff
@@ -0,0 +1,27 @@
+name: build_docs
```
Maybe we can align a bit more with the worfklow used in pynxtools: https://github.com/FAIRmat-NFDI/pynxtools/blob/master/.github/workflows/build_docs.yml

RubelMozumder replied:
Addressed at 53d263ed5edd4ec3ca20fb55b07674980f8d137e.

RubelMozumder replied:
Addressed at 53d263ed5edd4ec3ca20fb55b07674980f8d137e

lukaspie commented on README.md:
```diff
@@ -22,536 +22,7 @@
 8. [Conclusion](#conclusion)
 9. [References](#references)
```
please remove this table of contents as well

RubelMozumder replied:
Addressed at 53d263ed5edd4ec3ca20fb55b07674980f8d137e

lukaspie commented on docs/index.md:
```diff
@@ -0,0 +1,53 @@
+# FAIRmat nyaml documentation
+The NeXus data format, described by the NeXus Definition Language (NXDL), represents a concerted effort aimed at facilitating data exchange within scientific communities, particularly among those engaged in neutron, X-ray, and muon research [J. Appl. Cryst. (2015). 48, 301-305](https://doi.org/10.1107/S1600576714027575). The data format is also being used by the material science community under t ...
```
Now that I see it here, maybe this is actually too much for the general index. If we build the docs, we have to scroll a lot to see the different tiles for the individual parts of the plugin. My bad!

My suggestion:
- for the index page, go with a smaller description of the tool (like we do in the NOMAD [docs](https://nomad-lab.eu/prod/v1/staging/docs/))
- Whatever does fit into this short introduction, move it back to the explanations.
- For this, I suggest that we have two pages in the explanations part:
  - One is an actual explanation of what the tool is for (i.e., the long text we now have in index.md) -> goes to explanations.md
  - The other is what we already have in the explanations section. This is concerned with the workflow, so I would move it to a new file called workflow.md
- Of course, the links in index.md and mkdocs.yml need to be updated afterwards.

Sounds okay?

RubelMozumder replied:
Addressed at 53d263ed5edd4ec3ca20fb55b07674980f8d137e

lukaspie commented on docs/index.md:
```diff
@@ -0,0 +1,53 @@
+
+- [An introduction to NeXus](https://manual.nexusformat.org/index.html)
+- [Explanation of the nyaml tool](explanations.md)
+
```
Add a third page here for the workflow, as described above

lukaspie commented on docs/references.md:
```diff
@@ -0,0 +1,15 @@
+List of references used in this documentation:
+
+1. <a name="nyaml GitHub repository"></a>[nyaml GitHub repository](https://github.com/FAIRmat-NFDI/nyaml)
+2. <a name="NeXus-FAIRmat"></a>[NeXus-FAIRmat](https://fairmat-experimental.github.io/nexus-fairmat-proposal/)
```
```suggestion
2. <a name="NeXus-FAIRmat proposal"></a>[NeXus-FAIRmat proposal](https://fairmat-experimental.github.io/nexus-fairmat-proposal/)
```

RubelMozumder replied:
Added in [5ec2c7b](https://github.com/FAIRmat-NFDI/nyaml/commit/5ec2c7b07b738aac1a483248431b410e0c8577f8)

lukaspie commented on docs/tutorials.md:
```diff
@@ -0,0 +1,199 @@
+# Tutorials
```
Use a more descriptive title here. Tutorial for what?

RubelMozumder replied:
Addressed at 53d263ed5edd4ec3ca20fb55b07674980f8d137e

lukaspie commented on docs/tutorials.md:
```diff
@@ -0,0 +1,199 @@
+## Design of NeXus Ontology and Terms in YAML
+
+Within the YAML format, the root section denotes the top-level description of the application definition or base class schema, comprising the `category`, `type`, `doc`, `symbols` block, and the name of the schema (e.g. `NXmpes(NXobject)`). Correspondingly, the root section refers to the XML element `definition`, encompassing the first `doc` child of the `definition` and `symbols`. The definition  ...
+Within the YAML format, the root section denotes the top-level description of the application definition or base class schema, comprising the `category`, `type`, `doc`, `symbols` block, and the name of the schema (e.g. `NXmpes(NXobject)`). Correspondingly, the root section refers to the XML element `definition`, encompassing the first `doc` child of the `definition` and `symbols`. The definition  ...
```
Add line breaks, now it looks like a very big bloc in the built docs.

![image](https://github.com/user-attachments/assets/f0c46b8b-1988-4e84-8d0e-5564b3f3d25e)

lukaspie commented on docs/tutorials.md:
```diff
@@ -0,0 +1,199 @@
+
+### NeXus Group
+[NeXus groups](https://manual.nexusformat.org/design.html#design-groups), as instances of NeXus base classes, embody the compositional structure of application definitions. Depending on the value of keyword `nameType`, these groups can be initialized dynamically or statically, each approach offering distinct advantages. The keyword `nameType` can hold three district values: `specified`, `any`, an ...
+
```
Here it is better with the linebreaks.

lukaspie commented on docs/tutorials.md:
```diff
@@ -0,0 +1,199 @@
+
+Dynamic initialization allows the instantiation of groups while implementing the NeXus definition to store data (in HDF5 file format called NeXus file). This method provides flexibility for multiple instances at the same level within the NeXus file. For instance, the group `(NXmanipulator)` (with `nameType`=`any`) can initialize multiple groups such as `manipulator1` and `manipulator2` of the bas ...
+
+Descriptive information about NeXus groups is encapsulated within the `doc` child of the respective group. It is important to note that the group annotation of `source_TYPE(NXsource)` or `(NXsource)source_TYPE` signifies the encapsulation of the group's `name` as `source_TYPE` and its type as `NXsource` base class. Notably, the order between `name` and `type` within the XML element will be invert ...
```
I think we are enforcing `source_TYPE(NXsource)` now, so this part can be dropped

RubelMozumder replied:
Yeah, but I do not give any stress on any specific way of dynamic innitialisation, both `(nx_class)` and `static_nameDANYMICNAMNE`, both should come into the picture, no?

lukaspie replied:
My point is that we never write `(NXsource)source_TYPE` anymore, so we should not mention it herr

RubelMozumder replied:
Yeah, we do not write a group in that way if we develop from scratch. It is still writeable and indeed we get such syntax while extending or converting other nxdl file to yaml file developed by external people.

It is not anything wrong, right? IMO, by removing that we mainly hiding the another way of group representation in yaml which may be appeared in several app def or base class.

you may rethink that, I can remove that part, no problem.

RubelMozumder replied:
Addressed in 79eabd68751c64902af81894c0d684b2a8c280d3.

lukaspie commented on docs/tutorials.md:
````diff
@@ -0,0 +1,199 @@
+```
+
+### NeXus Field and NeXus Attribute
+A NeXus group may contain NeXus fields, NeXus attributes, and NeXus groups. A field, that does not have preceding `NX`, and a attribute, preceded by `\@`, must have a [NeXus type](https://manual.nexusformat.org/nxdl-types.html#index-0) (e.g.`NX_FLOAT`, `NX_CHAR`). In the YAML notation, each NeXus field or attribute has a implicit type `NX_CHAR`; otherwise its type must be denoted inside the paren ...
````
```suggestion
A NeXus group may contain NeXus fields, NeXus attributes, and other NeXus groups. A field, which is written as a string without a preceding `NX`, and an attribute, preceded by `\@`, must have a [NeXus type](https://manual.nexusformat.org/nxdl-types.html#index-0) (e.g.`NX_FLOAT`, `NX_CHAR`). The NeXus type type must be denoted inside parenthesis (e.g. `end_time(NX_DATE_TIME)`; if it is ommited, the NeXus field or attribute has the type `NX_CHAR` by default. Other XML attributes of the NeXus `field`/`attribute` can be defined using one of a set of special key ords below the field/attribute by indentation (the special keywords will be discussed in the next section). The explanatory documentation text of the NeXus field/attribute is given within the `doc` child.
```

Also make sure to either always use `field`/`attribute` (i.e., code-like) or just plain field/attribute.

mkuehbach replied:
There is here for me too much giving references to things and overall too much stuff in that flat file, better make tutorial.md have an own table of concepts so that one can navigate from section to section, also making reference to specific locations in the NeXus manual we should use whenever possible

RubelMozumder replied:
Added in [5ec2c7b](https://github.com/FAIRmat-NFDI/nyaml/commit/5ec2c7b07b738aac1a483248431b410e0c8577f8)

lukaspie commented on docs/tutorials.md:
```diff
@@ -0,0 +1,199 @@
+### NeXus Field and NeXus Attribute
+A NeXus group may contain NeXus fields, NeXus attributes, and NeXus groups. A field, that does not have preceding `NX`, and a attribute, preceded by `\@`, must have a [NeXus type](https://manual.nexusformat.org/nxdl-types.html#index-0) (e.g.`NX_FLOAT`, `NX_CHAR`). In the YAML notation, each NeXus field or attribute has a implicit type `NX_CHAR`; otherwise its type must be denoted inside the paren ...
+
+A NeXus group may encompass multiple `field`, `attribute`, and subgroup, each serving distinct purposes within the data structure. The [`field`](https://manual.nexusformat.org/design.html#design-fields), denoted without the prefix NX, and the [`attribute`](https://manual.nexusformat.org/design.html#design-attributes), indicated by `\@`, must be associated with a NeXus type (e.g., `NX_FLOAT`, `NX_ ...
```
This can be removed, it is a repetition of the paragraph before.

RubelMozumder replied:
Added in [5ec2c7b](https://github.com/FAIRmat-NFDI/nyaml/commit/5ec2c7b07b738aac1a483248431b410e0c8577f8)

lukaspie commented on docs/tutorials.md:
```diff
@@ -0,0 +1,199 @@
+
+A NeXus group may encompass multiple `field`, `attribute`, and subgroup, each serving distinct purposes within the data structure. The [`field`](https://manual.nexusformat.org/design.html#design-fields), denoted without the prefix NX, and the [`attribute`](https://manual.nexusformat.org/design.html#design-attributes), indicated by `\@`, must be associated with a NeXus type (e.g., `NX_FLOAT`, `NX_ ...
+
+Additionally, `XML` attributes specific to NeXus field and attribute are represented as children of the corresponding `field` or `attribute` elements (further details on special keywords will be discussed in the following section). Descriptive information pertaining to NeXus `field`s or `attribute`s is encapsulated within the `doc` child element.
```
Remove as described above

RubelMozumder replied:
Addressed in [5ec2c7b](https://github.com/FAIRmat-NFDI/nyaml/commit/5ec2c7b07b738aac1a483248431b410e0c8577f8)

lukaspie commented on docs/tutorials.md:
````diff
@@ -0,0 +1,199 @@
+```
+
+### NeXus Link
+The NeXus `link` concept reduces duplication of the data since several concepts of the same kind (e.g., NeXus field or NeXus attribute) can refer to a single copy of a data element. In YAML format, NeXus `link` is defined denoting the link in side parenthesis. At the same time, the concept containing the data must be mentioned under the `target` child.
````
```suggestion
The NeXus `link` reduces data duplication since several concepts of the same kind (e.g., NeXus group, field, or attribute) can refer to a single copy of a data element. In YAML format, NeXus `link` is defined denoting the link in parenthesis. At the same time, the concept containing the data must be mentioned under the `target` child.
```

RubelMozumder replied:
Addressed in [5ec2c7b](https://github.com/FAIRmat-NFDI/nyaml/commit/5ec2c7b07b738aac1a483248431b410e0c8577f8)

lukaspie commented on docs/tutorials.md:
````diff
@@ -0,0 +1,199 @@
+  doc: A link to a full data collection.
+```
+
+In the above YAML example, `reference_measurement` is defined as a link refering the `NXentry` group with its target specified as `/entry`. This structure ensures that the concept referencing the data is effectively linked to the designated target, thereby reducing redundancy and maintaining data integrity within the NeXus framework.
````
```suggestion
In the YAML example above, `reference_measurement` is defined as a link refering to the `NXentry` group with its target specified as `/entry`. This structure ensures that the concept referencing the data is effectively linked to the designated target, thereby reducing redundancy and maintaining data integrity within the NeXus framework.
```

RubelMozumder replied:
Addressed in [5ec2c7b](https://github.com/FAIRmat-NFDI/nyaml/commit/5ec2c7b07b738aac1a483248431b410e0c8577f8)

lukaspie commented on docs/tutorials.md:
```diff
@@ -0,0 +1,199 @@
+In this `choice` example, `pixel_shape` could be any of the groups `(NXoff_geometry)` and `(NXcylindrical_geometry)`, depending on the geometry of the pixels. -->
+
+## Special Keywords in YAML
+To explain the context of NeXus, certain keywords hold significance beyond their literal interpretations. These special keywords are utilized to elucidate and denote various NeXus terms like attributes, fields, links, and groups, thereby enhancing the clarity and specificity of the data representation.
```
```suggestion
In the YAML schema the tool uses, certain keywords hold significance beyond their literal interpretations. These special keywords are utilized to elucidate and denote various NeXus terms like attributes, fields, links, and groups, thereby enhancing the clarity and specificity of the data representation.
```

RubelMozumder replied:
Addressed in [5ec2c7b](https://github.com/FAIRmat-NFDI/nyaml/commit/5ec2c7b07b738aac1a483248431b410e0c8577f8)

lukaspie commented on docs/tutorials.md:
```diff
@@ -0,0 +1,199 @@
+To explain the context of NeXus, certain keywords hold significance beyond their literal interpretations. These special keywords are utilized to elucidate and denote various NeXus terms like attributes, fields, links, and groups, thereby enhancing the clarity and specificity of the data representation.
+
+### Keyword `exists`
+The `exists` keyword plays a pivotal role in delineating the optionality of NeXus concepts `attribute`, `field`, `choice` `link`, and `group`, during the implementation of NeXus definitions in NeXus files. It provides crucial insights into the expected presence or absence of these concepts within the NeXus data structure. By default, all the concepts of a base class are optional, while in an appl ...
```
remove choice here

RubelMozumder replied:
commented out.

lukaspie commented on docs/tutorials.md:
```diff
@@ -0,0 +1,199 @@
+To explain the context of NeXus, certain keywords hold significance beyond their literal interpretations. These special keywords are utilized to elucidate and denote various NeXus terms like attributes, fields, links, and groups, thereby enhancing the clarity and specificity of the data representation.
+
+### Keyword `exists`
+The `exists` keyword plays a pivotal role in delineating the optionality of NeXus concepts `attribute`, `field`, `choice` `link`, and `group`, during the implementation of NeXus definitions in NeXus files. It provides crucial insights into the expected presence or absence of these concepts within the NeXus data structure. By default, all the concepts of a base class are optional, while in an appl ...
```
```suggestion
The `exists` keyword plays a pivotal role in definining the optionality of NeXus concepts `attribute`, `field`, `link`, and `group`. It provides crucial insights into the expected presence or absence of these concepts within the NeXus data structure when implementing the NeXus definitions in NeXus files. By default, all the concepts of a base class are optional, while in an application definition, all concepts are required.
```

RubelMozumder replied:
Addressed in [5ec2c7b](https://github.com/FAIRmat-NFDI/nyaml/commit/5ec2c7b07b738aac1a483248431b410e0c8577f8)

lukaspie commented on docs/tutorials.md:
```diff
@@ -0,0 +1,199 @@
+
+`optional`: Denotes that the NeXus concept is not mandatory and may be absent.
+`recommended`: Suggests that the NeXus concept is advisable, but not mandatory.
+`required`: Indicates that the NeXus concept must be present within the structure. Any validation of a NeXus file will fail or give warning if required concepts (for a given application definition) are not available.
```
```suggestion
`required`: Indicates that the NeXus concept must be present within the structure. Any validation of a NeXus file will fail or give a warning if required concepts (for a given application definition) are not available.
```

RubelMozumder replied:
Addressed in [5ec2c7b](https://github.com/FAIRmat-NFDI/nyaml/commit/5ec2c7b07b738aac1a483248431b410e0c8577f8)

lukaspie commented on docs/tutorials.md:
````diff
@@ -0,0 +1,199 @@
+  doc: |
+    This calibration procedure is used to account for the different tranmsission efficiencies.
+```
+In the above example the greoup `transmission_correction` is a optional group.
````
```suggestion
In the above example the group `transmission_correction` is an optional group.
```

lukaspie commented on docs/tutorials.md:
```diff
@@ -0,0 +1,199 @@
+In the above example the greoup `transmission_correction` is a optional group.
+
+### Keyword `unit`
+A statement introducing NeXus-compliant NXDL `units` attribute to the `field`, e.g. `NX_VOLTAGE` to assign a predefined physical unit.
```
```suggestion
The `unit` keyword is used to define the NeXus-compliant unit categories (which int the NXDL schema are denoted as `units`).
```

RubelMozumder replied:
Addressed in [5ec2c7b](https://github.com/FAIRmat-NFDI/nyaml/commit/5ec2c7b07b738aac1a483248431b410e0c8577f8)

lukaspie commented on docs/tutorials.md:
````diff
@@ -0,0 +1,199 @@
+```
+
+### Keyword `dimensions`
+The `dimensions` term  describes the multidimensional nature of the data, specifying its rank, dimensional indices, and corresponding length of the rank. For example, the attribute `rank` defines the dimension of the data set. To elucidate each dimension, we use two other keywords: `dim` and `dim_parameters`. The `dim` keyword comprises an array of arrays, the nested array encapsulates values for ...
````
```suggestion
The `dimensions` term  describes the multidimensional nature of the data, specifying its rank, dimensional indices, and corresponding length of the rank. The attribute `rank` defines the dimension of the data set. To elucidate each dimension, we use two other keywords: `dim` and `dim_parameters`. The `dim` keyword comprises an array of arrays, the nested array encapsulates values for `index` and `value` (NeXus keywords) pairs. Each array within the `dim` array corresponds to a specific dimension of the multidimensional data. For example, for 2D particle motion, the `dim` array may be represented as `[[0, nx], [1, ny]]`, each member indicating the axis index and axis name. The keyword `dim_parameters` contains further information of each dimension such as `doc`, `ref`, etc. It is important to note that each term or keyword within `dim_parameters` must have the same length as the value of the rank keyword.
```

This reads a bit confusing, but probably okay

RubelMozumder replied:
Added in [5ec2c7b](https://github.com/FAIRmat-NFDI/nyaml/commit/5ec2c7b07b738aac1a483248431b410e0c8577f8)

lukaspie commented on docs/tutorials.md:
````diff
@@ -0,0 +1,199 @@
+```
+
+### Keyword `enumeration`
+Python-like list of strings which are considered as recommended items for the fields or attributes.
````
```suggestion
List of strings which are considered as a set of predefined values for fields or attributes.
```

lukaspie replied:
Add something about open enums here.

lukaspie commented on docs/tutorials.md:
````diff
@@ -0,0 +1,199 @@
+In the example, the only valid value for NeXus field `definition` is `NXmpes`.
+```
+In the example, the only valid value for NeXus field `definition` is `NXmpes`.
+The `xref` keyword (which can only be used inside the keyword `doc`) is used to refer any other ontology or any other standard such `ISO`. The `xref` in the example `doc` will reflect the information inside the XML `doc`. Note that the `xref` keyword is only available in the `YAML` representation and will be transformed into its textual representation inside the `doc` text in `XML`.
````
Remove the part starting from "The `xref` keyword", this is mentioned in the next section

RubelMozumder replied:
Added in [5ec2c7b](https://github.com/FAIRmat-NFDI/nyaml/commit/5ec2c7b07b738aac1a483248431b410e0c8577f8)

lukaspie commented on docs/tutorials.md:
```diff
@@ -0,0 +1,199 @@
+In the example, the only valid value for NeXus field `definition` is `NXmpes`.
+The `xref` keyword (which can only be used inside the keyword `doc`) is used to refer any other ontology or any other standard such `ISO`. The `xref` in the example `doc` will reflect the information inside the XML `doc`. Note that the `xref` keyword is only available in the `YAML` representation and will be transformed into its textual representation inside the `doc` text in `XML`.
+### Keyword `xref`
+The `xref` keyword (which can only be used inside the keyword `doc`) is used to refer any other ontology or any other standard such `ISO`. The `xref` in the example `doc` will reflect the information inside the XML `doc`. Note that the `xref` keyword is only available in the `YAML` representation and will be transformed into its textual representation inside the `doc` text in `XML`.
```
```suggestion
The `xref` keyword (which can only be used inside the keyword `doc`) is used to refer any other ontology or any other standard (such as `ISO`). The `xref` in the example `doc` will reflect the information inside the XML `doc`. Note that the `xref` keyword is only available in the `YAML` representation and will be transformed into its textual representation inside the `doc` text in `XML`.
```

lukaspie commented on docs/tutorials.md:
````diff
@@ -0,0 +1,199 @@
+      url: https://www.iso.org/obp/ui/en/#iso:std:iso:18115:-1:ed-3:v1:en:term:12.58
+```
+
+### Keyword `nameType`
````
I would push this up, right to where we are discussing the name and the replacement if nameType is partial or any

RubelMozumder committed:
rebase.

mkuehbach commented on docs/explanations.md:
```diff
@@ -0,0 +1,212 @@
+## nyaml Workflow
+For a given input file, the `nyaml` converter checks for the correct file type and call appropriate converter. For an XML file, the `nxdl2yaml` converter parses the `XML` file, by means of [lxml](https://lxml.de/) python library, into an `XML` tree object. Adhering to the NXDL rules, the converter writes the application definition or the base class object to a `yaml` file that matches the `nyaml` ...
```
"Provided an input file, the `nyaml` converter ..... and calls one of two possible converters: For an `nxdl.xml` file this is the `nxdl2yaml` converter. For a `yaml`file this is the `yaml2nxdl` converter.

for the sentence with lxml reformulate "The lxml Python library (add proper ref) is used for parsing XML files into an `XML` tree object in memory."

RubelMozumder replied:
Tailored!

mkuehbach commented on docs/explanations.md:
```diff
@@ -0,0 +1,212 @@
+## nyaml Workflow
+For a given input file, the `nyaml` converter checks for the correct file type and call appropriate converter. For an XML file, the `nxdl2yaml` converter parses the `XML` file, by means of [lxml](https://lxml.de/) python library, into an `XML` tree object. Adhering to the NXDL rules, the converter writes the application definition or the base class object to a `yaml` file that matches the `nyaml` ...
+
+Like every scientific software, the `nyaml` tool also follows a specific workflow.
```
reformulate and shorten: "`nyaml` implements a specific workflow:"

mkuehbach commented on docs/explanations.md:
````diff
@@ -0,0 +1,212 @@
+
+Like every scientific software, the `nyaml` tool also follows a specific workflow.
+
+```mermaid
````
If we want such mermaid plot in the JOSS CI/CD?
If not, we should export a vector graphic beforehand.

lukaspie replied:
We need to check that it builds correctly using the mkdocs CI/CD, right? JOSS CI/CD is just for the paper draft.

That being said, in the current docs, it looks like this
![image](https://github.com/user-attachments/assets/49030c1b-f961-4dd0-9b74-22186dc431ed)

Is that how it's supposed to look? If not, converting to a graphic may be nicer.

mkuehbach commented on docs/explanations.md:
````diff
@@ -0,0 +1,212 @@
+```
+
+## Conversion from YAML to XML and vice versa
+Presented below is a concise and trimmed example of the `NXmpes` application definition (for discussion purpose) in `YAML` format, alongside its corresponding encoding into `XML` format, as illustrated below. Subsequently, the fundamental rules governing this conversion process are elucidated. For a comprehensive understanding of the basic structure of NXDL, readers are encouraged to explore the  ...
````
The part "NOTE: For in" should be moved to a highlighting box that are used also in other places and plugins in NOMAD

mkuehbach replied:
this is a part in the learning section following the diataxis paradigm I do not really get why at all here a full excerpt of an application definition is shown, if we show one show a shorter one that has the complexities in it but is not as long, 
but also nothing is here really explained just the result of the nxdl to yaml conversion is presented without substantiated explanation.

mkuehbach commented on docs/explanations.md:
```diff
@@ -0,0 +1,212 @@
+    Number of data points in the transmission function.
+NXmpes(NXobject):
+  (NXentry):
+    exsits: required
```
this is for sure wrong

mkuehbach replied:
we must not share incorrect explanations!

RubelMozumder replied:
Addresed!

mkuehbach commented on docs/explanations.md:
```diff
@@ -0,0 +1,212 @@
+  (NXentry):
+    exsits: required
+    definition:
+      \@version:
```
this as well intendation

mkuehbach commented on docs/explanations.md:
```diff
@@ -0,0 +1,212 @@
+    definition:
+      \@version:
+      enumeration: [NXmpes]
+    title:
```
having on example like

a_char_concept(NX_CHAR)

and another one with

another_char_concept:

is useful indeed in an explanation if complemented with the explanation that because NX_CHAR is assumed the default writing NX_CHAR can be omitted. This is though the only type that can be omitted, which should also be stated.

mkuehbach commented on docs/explanations.md:
```diff
@@ -0,0 +1,212 @@
+      - |
+        Description of the MPES spectrometer and its individual parts.
+      - |
+        xref:
```
this for sure substantiates to have an explanation

my proposal is to split the individual snippets of the appdef apart and have one code snippet per feature but then this qualifies for moving it into the references.md

mkuehbach commented on docs/explanations.md:
```diff
@@ -0,0 +1,212 @@
+          spec: ISO 18115-1:2023
+          term: 12.58
+          url: https://www.iso.org/obp/ui/en/#iso:std:iso:18115:-1:ed-3:v1:en:term:12.58
+      source_TYPE(NXsource):
```
this for sure warrants explanation given that we have now nameType any capitalization is no longer decisive when it comes to defining if a concept name has variadic parts or not

RubelMozumder replied:
Explained with nameType.

mkuehbach commented on docs/explanations.md:
````diff
@@ -0,0 +1,212 @@
+            dimensions:
+              rank: 1
+              dim: [[1, n_transmission_function]]
+```
````
this should be splitted

mkuehbach commented on docs/explanations.md:
````diff
@@ -0,0 +1,212 @@
+              rank: 1
+              dim: [[1, n_transmission_function]]
+```
+
````
the functionality of nyaml that there is a caching in the sense that the source XML is attached to the tail of an yaml file should be explained and ideally get a small example snippet, it is important to mention that a hashing is used for figuring out if content changed

mkuehbach commented on docs/how_to_guide.md:
```diff
@@ -0,0 +1,45 @@
+# nyaml Installation and Command
+`nyaml` is a Python package [published on PyPI](https://pypi.org/project/nyaml/).
```
"that is [published on PyPI]"

mkuehbach commented on docs/how_to_guide.md:
recommend to change filename to "howto.md"

RubelMozumder replied:
Addressed!

mkuehbach commented on docs/index.md:
```diff
@@ -0,0 +1,53 @@
+# FAIRmat nyaml documentation
+The NeXus data format, described by the NeXus Definition Language (NXDL), represents a concerted effort aimed at facilitating data exchange within scientific communities, particularly among those engaged in neutron, X-ray, and muon research [J. Appl. Cryst. (2015). 48, 301-305](https://doi.org/10.1107/S1600576714027575). The data format is also being used by the material science community under t ...
```
"what does the "It" in "data principles. It" refer to?

RubelMozumder replied:
Addresed!

mkuehbach commented on docs/index.md:
```diff
@@ -0,0 +1,53 @@
+# FAIRmat nyaml documentation
+The NeXus data format, described by the NeXus Definition Language (NXDL), represents a concerted effort aimed at facilitating data exchange within scientific communities, particularly among those engaged in neutron, X-ray, and muon research [J. Appl. Cryst. (2015). 48, 301-305](https://doi.org/10.1107/S1600576714027575). The data format is also being used by the material science community under t ...
+
+NXDL is used to define general data storage objects (base classes) and the base classes are the building blocks for defining measurement-specific or even instrument-specific or software-specific data storage objects (application definitions). In this process, members and definitions of individual base classes can be used as is or customized. In essence, the process of schema development, whether  ...
```
Adding to Lukas' recommendation I suggest to do the splitting first and then start a revisioning of the text paragraph snippets remaining, for this reason I skip the revision here for this file.

mkuehbach commented on docs/references.md:
```diff
@@ -0,0 +1,15 @@
+1. <a name="nyaml GitHub repository"></a>[nyaml GitHub repository](https://github.com/FAIRmat-NFDI/nyaml)
+2. <a name="NeXus-FAIRmat"></a>[NeXus-FAIRmat](https://fairmat-experimental.github.io/nexus-fairmat-proposal/)
+3. <a name="FAIRmat-NeXus GitHub repository"></a>[FAIRmat-NeXus GitHub repository](https://github.com/FAIRmat-NFDI/nexus_definitions)
+4. <a name="official site"></a>[Official NeXus website](https://manual.nexusformat.org/index.html)
```
"official site" > "Official NeXus website"

mkuehbach commented on docs/references.md:
```diff
@@ -0,0 +1,15 @@
+2. <a name="NeXus-FAIRmat"></a>[NeXus-FAIRmat](https://fairmat-experimental.github.io/nexus-fairmat-proposal/)
+3. <a name="FAIRmat-NeXus GitHub repository"></a>[FAIRmat-NeXus GitHub repository](https://github.com/FAIRmat-NFDI/nexus_definitions)
+4. <a name="official site"></a>[Official NeXus website](https://manual.nexusformat.org/index.html)
+5. <a name="j_appl_cryst_2015"></a>[J. Appl. Cryst. (2015). 48, 301-305](https://doi.org/10.1107/S1600576714027575)
```
"j_appl_cryst_2015" > "Könnecke, M. et al."

mkuehbach commented on docs/references.md:
```diff
@@ -0,0 +1,15 @@
+8. <a name="NeXus official website"></a>[NeXus official website](https://www.nexusformat.org/)
+9. <a name="YAML"></a>[YAML](https://yaml.org/)
+10. <a name="lxml"></a>[lxml](https://lxml.de/)
+11. <a name="NeXus Manual"></a>[NeXus Manual](https://manual.nexusformat.org/user_manual.html)
```
we should check that all these links are still working

mkuehbach commented on docs/references.md:
```diff
@@ -0,0 +1,15 @@
+7. <a name="nyaml"></a>[nyaml](https://github.com/FAIRmat-NFDI/nyaml/tree/main)
+8. <a name="NeXus official website"></a>[NeXus official website](https://www.nexusformat.org/)
+9. <a name="YAML"></a>[YAML](https://yaml.org/)
+10. <a name="lxml"></a>[lxml](https://lxml.de/)
```
Follow these rules:
http://citebay.com/how-to-cite/lxml/

RubelMozumder replied:
This is not a citation as goes with publication. But a general reference link used software or information. Or? But the such citation must come in Joss paper as you mentioned in here.

mkuehbach commented on docs/stylesheets/extra.css:
does this implement the latest corporate design which e.g. pynxtools is using as well, if not, make a specific PR for changing this that can be merged into src branch of this PR

mkuehbach commented on docs/tutorials.md:
```diff
@@ -0,0 +1,199 @@
+# Tutorials
+This explanatory tutorial will explain different keywords, terms, and rules from the perspective of YAML format of the NeXus schema. Therefore it provides a overall grasp on how to write a NeXus schema those syntactical components.
```
Paragraph needs to be completely rewritten.

On the one hand for grammar and language.
On the other hand for precision of the formulations,
a NeXus data schema's XML schema may have a serialization in YAML but other than this there is no connection between NeXus and YAML.

mkuehbach commented on docs/tutorials.md:
```diff
@@ -0,0 +1,199 @@
+
+## Design of NeXus Ontology and Terms in YAML
+
+Within the YAML format, the root section denotes the top-level description of the application definition or base class schema, comprising the `category`, `type`, `doc`, `symbols` block, and the name of the schema (e.g. `NXmpes(NXobject)`). Correspondingly, the root section refers to the XML element `definition`, encompassing the first `doc` child of the `definition` and `symbols`. The definition  ...
```
This paragraph is an enumeration of mappings between where something is in XML and where it is in yaml and vice versa, the content should be splitted in shorter paragraphs, statements about what is in the top-level section should go to a paragraph on top-level section, statements for a appdefs going into one in appdefs.

This is a tutorial, meaning taking people by the hand, a wall of mappings provides no taking by some hand, currently not enough bandwidth to edit this on the train

maybe its an approach to do first the moving of content like suggested by Lukas and then make individual follow-up PRs e.g. one for each md file, explanations, howto,
this would enable a more focused "head writing" and makes the reviewing shorter instead of trying to lump all changes in one PR

mkuehbach reviewed (changes requested):
Implement the suggestions from Lukas first, especially the splitting and moving of content, apply then my comments on top

RubelMozumder committed:
PR update from Lukas comment.

PR change request.

RubelMozumder commented:
Do a final check if all links work!

RubelMozumder committed:
remove discussion between the (base_class)instance_name and instance_name(base_class).

lukaspie commented:
Thanks for splitting the explanations and workflow parts. I think for editability in the future it may be good to have separate folders for the 4 diataxis groups (tutorial, learn, etc.) into which we place the respective files.

lukaspie commented on docs/explanations.md:
```diff
@@ -0,0 +1,8 @@
+# NeXus Definition Language (NXDL), XML, and YAML
```
Would move the "At its core, the NeXus Definition Language (NXDL)..." sentence to the start of the next paragraph

RubelMozumder replied:
Addressed!

lukaspie commented on docs/explanations.md:
```diff
@@ -0,0 +1,8 @@
+# NeXus Definition Language (NXDL), XML, and YAML
+The NeXus data format, described by the NeXus Definition Language (NXDL), represents a concerted effort aimed at facilitating data exchange within scientific communities, particularly among those engaged in neutron, X-ray, and muon research [J. Appl. Cryst. (2015). 48, 301-305](https://doi.org/10.1107/S1600576714027575). The data format is also being used by the material science community under t ...
+
+NXDL is used to define general data storage objects (base classes) and the base classes are the building blocks for defining, under the schema called application definition,  measurement-specific or even instrument-specific or software-specific data storage objects. In this process, members and definitions of individual base classes can be used as is or customized. In essence, the process of sche ...
```
```suggestion
NXDL is used to define general data structures (base classes), which define the set of terms that might be used in an instance of that class. These base classes are the building blocks for defining, under the schema called application definition, measurement-specific or even instrument-specific or software-specific data storage objects. In this process, members and definitions of individual base classes can be used as is or customized. In essence, the process of schema development, whether for a base class or an application definition, entails crafting an NXDL schema definition file with the extension 'nxdl.xml', utilizing the Extensible Markup Language, [XML](https://www.w3.org/TR/REC-xml/REC-xml-20081126.xml).
```

We do not define data storage objects!

RubelMozumder replied:
Addressed!

RubelMozumder committed:
tiny fix.

lukaspie commented on docs/explanations.md:
```diff
@@ -0,0 +1,8 @@
+
+NXDL is used to define general data storage objects (base classes) and the base classes are the building blocks for defining, under the schema called application definition,  measurement-specific or even instrument-specific or software-specific data storage objects. In this process, members and definitions of individual base classes can be used as is or customized. In essence, the process of sche ...
+
+To expedite the schema development process, we have introduced the use of Yet Another Markup Language ([YAML](https://yaml.org/)), which provides a syntax or style specifically tailored for defining scientific domain-driven schemas with NXDL. One significant advantage of YAML over XML is its indentation-driven approach, which eliminates the need for starting and ending tags for each entity within ...
```
```suggestion
To expedite the schema development process, we have introduced the use of Yet Another Markup Language ([YAML](https://yaml.org/)), which provides a syntax or style specifically tailored for defining scientific domain-driven schemas with NXDL. One significant advantage of YAML over XML is its indentation-driven approach, which eliminates the need for starting and ending tags for each entity within the schema. The YAML format reduces repetition of NXDL keywords and offers a more intuitive structure aligned with object-oriented programming concepts, such as class inheritance. These benefits are attained without compromising the integrity of the original NeXus schema, which is traditionally expressed in XML format.
```

always use either YAML or `YAML` (i.e., plain text or code)

RubelMozumder replied:
Addressed!

lukaspie commented on docs/explanations.md:
```diff
@@ -0,0 +1,8 @@
+
+To expedite the schema development process, we have introduced the use of Yet Another Markup Language ([YAML](https://yaml.org/)), which provides a syntax or style specifically tailored for defining scientific domain-driven schemas with NXDL. One significant advantage of YAML over XML is its indentation-driven approach, which eliminates the need for starting and ending tags for each entity within ...
+
+The `YAML` format, while not yet an official version of NeXus application definitions or base classes, necessitates a method for transcoding it into `XML`. The [nyaml](https://github.com/FAIRmat-NFDI/nyaml) Python package serves as a converter tool designed specifically for this purpose. It enables the conversion of NXDL from `YAML` format to `XML`, thereby enhancing the capability of NeXus schem ...
```
```suggestion
While `YAML` offers a more readable and concise way to author NeXus definitions, a transcoding mechanism is required because the official format remains XML. The [nyaml](https://github.com/FAIRmat-NFDI/nyaml) Python package serves as a converter tool designed specifically for this purpose. It enables the conversion of NXDL-compliant definitions from YAML to XML, making it easier for NeXus schema developers to create and maintain definitions. Furthermore, the tool offers the flexibility to extend existing NeXus schemas in XML by facilitating conversion back and forth between the two formats. It is important to note that here we do not introduce NeXus data objects, terms, or types, which are fundamental for writing base class schemas or application definition schemas. For individuals new to NeXus, we refer to the official NeXus site at NeXus ([NeXus official website](https://www.nexusformat.org/)).
```

RubelMozumder replied:
Addresseds!

lukaspie commented on docs/index.md:
```diff
@@ -0,0 +1,53 @@
+
+The `nyaml` tool is designed to be user-friendly, with a command line interface that allows users to easily inter-convert between `YAML` and `XML` formats written in terms of NXDL (NeXus Definition Language). This toolset is particularly beneficial for scientists and developers who are familiar with the NeXus format and wish to leverage the advantages of `YAML` for schema development. The tool su ...
+
+The tool can be installed via pip, and it provides comprehensive documentation and tutorials to assist users in writing NeXus definitions in `YAML`.
```
```suggestion
The `nyaml` tool offers a user-friendly command-line interface for converting between `YAML` and `XML` representations of NeXus definitions written in the NeXus Definition Language (NXDL). While the official NeXus definitions are maintained in `XML,` authoring them in `YAML` provides a more concise and readable alternative. `nyaml` bridges this gap, allowing developers to work in `YAML` while ensuring compatibility with the official `XML` format.

The tool supports both application definitions and base classes. It can be installed via `pip` and includes comprehensive documentation and tutorials to assist users in writing NeXus definitions using `YAML`.
```

We should mention what the package solves at the beginning.

RubelMozumder replied:
Addressed!

lukaspie commented on docs/index.md:
```diff
@@ -0,0 +1,53 @@
+
+How-to install and use the nyaml tool
+
+- [Use of `nyaml` tool](how_to_guide.md)
```
```suggestion
- [Installation and use of the `nyaml` tool](how_to_guide.md)
```

RubelMozumder replied:
Addressed!

lukaspie commented on docs/index.md:
```diff
@@ -0,0 +1,53 @@
+
+### Learn
+
+An introduction to NeXus and its design principles.
```
Remove this sentence, it does not match the individual pages

RubelMozumder replied:
Modified!

lukaspie reviewed (commented):
Couple more comments, already going in the right direction

RubelMozumder committed:
General map between the yaml and xml version of nxdl

RubelMozumder committed:
Update the docs in tuorials.md.

Update minor fix.

RubelMozumder committed:
Hash mechanism.

Update PR requests.

Update docs/index.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

Update docs/explanations.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

Update docs/explanations.md

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

RubelMozumder committed:
use folder structure.

RubelMozumder committed:
Update the PR requests.

RubelMozumder committed:
foder structure.

RubelMozumder committed:
use nomad recomended design.

lukaspie commented on mkdocs.yml:
```diff
@@ -50,5 +51,11 @@ theme:
+
+extra:
+  generator: false
+  # homepage: https://fairmat-nfdi.github.io/pynxtools/
```
should be nyaml

RubelMozumder replied:
fixed!

RubelMozumder committed:
Fix links.

RubelMozumder committed:
Fix the alignment of contents and redundant features from mkdocs.yml.

updates.

mkuehbach reviewed (approved):
All remaining conversations have been resolved, the PR is already quite long, the CI runs through, this PR should be merged @RubelMozumder

lukaspie commented on mkdocs.yml:
```diff
@@ -0,0 +1,67 @@
+  - Tutorial: tutorials/tutorials.md
+  - How-To-Guide : howtos/howto.md
+  - Explanation: learn/explanations.md
+  - Reference: references/references.md
```
This sections needs to be updated so that the new pages are included in the left-hand side navigation

RubelMozumder replied:
Addressed!

lukaspie commented on mkdocs.yml:
```diff
@@ -0,0 +1,67 @@
+site_name: NYAML Documentation
+site_description:
+  The documentation for the NYAML tool which is a converter tool to transcode NeXus schema from YAML to XML (.nxdl.xml).
```
```suggestion
  Documentation for nyaml: a converter tool to transcode NeXus schemas from YAML to XML (.nxdl.xml).
```

mkuehbach replied:
Eventually also here XML to YAML and vice-versa

RubelMozumder replied:
addressed!

lukaspie commented on pyproject.toml:
```diff
@@ -42,6 +42,13 @@ dev = [
     "types-pyyaml",
     "lxml-stubs",
     "uv>=0.4.10",
+    # Mkdocs and the related plugins
```
I suggest to make these their own `docs` extra, like in pynxtools

RubelMozumder replied:
addressed!

lukaspie commented on src/nyaml/comment_collector.py:
```diff
@@ -342,7 +342,7 @@ def get_comment_text_list(self) -> Union[List, str]:
 
 
-class YAMLComment(Comment):
+class YAMLComment(Comment):  # noqa: PLW1641
```
I beleive this is no longer neccessary

RubelMozumder replied:
Addressed!

lukaspie commented on README.md:
```diff
@@ -1,557 +1,7 @@
-[@Könnecke]: J. Appl. Cryst. (2015). 48, 301-305
-             (https://doi.org/10.1107/S1600576714027575)
-NeXus-FAIRmat: https://fairmat-nfdi.github.io/nexus_definitions/
+Check full documentation of the tool in [gh] https://fairmat-nfdi.github.io/nyaml/
```
```suggestion
You can find the full documentation of the tool [here](https://fairmat-nfdi.github.io/nyaml/)
```
gh does not mean anything to new users

RubelMozumder replied:
Addressed!

lukaspie commented on docs/assets/.gitignore:
```diff
@@ -0,0 +1 @@
+nomad-oasis*.zip
```
Is this needed?

mkuehbach replied:
Seems some debris left from hotfixing in the past, so no

RubelMozumder replied:
removed!

lukaspie commented on mkdocs.yml:
```diff
@@ -0,0 +1,67 @@
+    accent: "#008A67"
+  font:
+    text: "Titillium Web"
+  logo: assets/nomad-plugin-logo.png
```
This is not a NOMAD plugin, so we should not use the NOMAD plugin logo. Can we just use a regular NOMAD logo? If not, I suggest we don't use any logo at all.

mkuehbach replied:
Agree, see suggestions made for the favicon and nomad-plugin-logo changes abovementioned

RubelMozumder replied:
New logos are added!

lukaspie commented on mkdocs.yml:
```diff
@@ -0,0 +1,67 @@
+      permalink: True
+  - pymdownx.arithmatex:
+      generic: true
+  - pymdownx.emoji: # TODO make sure this isn't used!!
```
Can you double check that this is needed for the docs to build?

RubelMozumder replied:
addressed!

lukaspie commented on docs/references/references.md:
```diff
@@ -0,0 +1,15 @@
+#References
```
```suggestion
# References
```

RubelMozumder replied:
Addressed!

lukaspie commented on docs/learn/workflow.md:
```diff
@@ -0,0 +1,222 @@
+## nyaml Workflow
+The `nyaml` implements a specific workflow for converting between YAML and XML formats. The `nyaml` converter checks for the correct file type and calls the appropriate converter. For an XML file, the `nxdl2yaml` converter parses the `XML` file, by means of [lxml](https://lxml.de/) python library, into an `XML` tree object. Adhering to the NXDL rules, the converter writes the application definiti ...
```
```suggestion
`nyaml` implements a specific workflow for converting between YAML and XML formats. The `nyaml` converter checks for the correct file type and calls the appropriate converter. For an XML file, the `nxdl2yaml` converter parses the `XML` file, by means of the [lxml](https://lxml.de/) Python library, into an `XML` tree object. Adhering to the NXDL rules, the converter writes the application definition or the base class object to a `YAML` file that matches the `nyaml` syntax. If the input file is a `YAML` file, the `yaml2nxdl` converter collects the comments in a `Comments` object and parses the `YAML` file into a python `dictionary` object. Afterwards, the application definitions or base classes is converted into an `XML` file by combining the `Comments` and the Python `dictionary` object.
```

mkuehbach replied:
"the application definitions or base classes" use singular from like for XML direction aforementioned

RubelMozumder replied:
Addressed!

lukaspie commented on docs/learn/workflow.md:
```diff
@@ -0,0 +1,222 @@
+## nyaml Workflow
+The `nyaml` implements a specific workflow for converting between YAML and XML formats. The `nyaml` converter checks for the correct file type and calls the appropriate converter. For an XML file, the `nxdl2yaml` converter parses the `XML` file, by means of [lxml](https://lxml.de/) python library, into an `XML` tree object. Adhering to the NXDL rules, the converter writes the application definiti ...
+
+After the conversion from `XML` format to `YAML` format, the `nyaml` tool stores the `XML` contents at the end of the file under a hash created by Secure Hash Algorithm sha256 from `YAML` contents generated in this conversion process. This benefits on the backward compatibility of the `YAML` file, so that the `YAML` file can be converted back to `XML` file without any modification of structure li ...
```
```suggestion
After the conversion from `XML` format to `YAML` format, the `nyaml` tool stores the `XML` contents at the end of the file under a hash created by Secure Hash Algorithm sha256 from `YAML` contents generated in this conversion process. This benefits the backward compatibility of the `YAML` file, so that the `YAML` file can be converted back to `XML` file without any modification of structure like whitespaces or newlines.
```

mkuehbach replied:
Rather a comment:

We should at least in one place though mention that this hashing is also used because of the fact that the round trip conversions do not contain order otherwise even if the tool were implemented more rigorously, the xsd tells clearly that complex types are sequences meaning the order of the XML attributes and concepts matters. For a normal set of concepts it does not. https://www.balisage.net/Proceedings/vol26/html/LaFontaine01/BalisageVol26-LaFontaine01.html and https://stackoverflow.com/questions/14917943/python-lxml-enforcing-a-specific-order-for-attributes where doing neither and caching is a practical workaround

https://lxml.de/FAQ.html
https://github.com/nexusformat/definitions/blob/8237a9c2c417d4d3d25f72c9e5e69a6707fca337/nxdl.xsd#L986

RubelMozumder replied:
Text for workflow explanation reviewed and updated!

lukaspie commented on docs/learn/workflow.md:
````diff
@@ -0,0 +1,222 @@
+```
+
+## Conversion from YAML to XML and vice versa
+Presented below is a concise and trimmed example of the `NXmpes` application definition (for discussion purpose) in `YAML` format, alongside its corresponding encoding into `XML` format, as illustrated below. Subsequently, the fundamental rules governing this conversion process are elucidated. For a comprehensive understanding of the basic structure of NXDL, readers are encouraged to explore the  ...
````
Can we link to NXmpes here?

mkuehbach replied:
The point about that NXmpes is picked here which is mentioned later that it is only an example should be pulled up here already

RubelMozumder replied:
Inlcuded in note!

lukaspie commented on docs/learn/workflow.md:
```diff
@@ -0,0 +1,222 @@
+Presented below is a concise and trimmed example of the `NXmpes` application definition (for discussion purpose) in `YAML` format, alongside its corresponding encoding into `XML` format, as illustrated below. Subsequently, the fundamental rules governing this conversion process are elucidated. For a comprehensive understanding of the basic structure of NXDL, readers are encouraged to explore the  ...
+
+!!! note
+    For indetailed explanation with examples please follow the [Tutorials for writing NeXus definition in YAML](../tutorials/tutorials.md).
```
```suggestion
    For a detailed explanation with examples please follow the [tutorial for writing NeXus definition in YAML](../tutorials/tutorials.md).
```

RubelMozumder replied:
Refactored the description of the workflow.

lukaspie commented on docs/learn/workflow.md:
```diff
@@ -0,0 +1,222 @@
+!!! note
+    For indetailed explanation with examples please follow the [Tutorials for writing NeXus definition in YAML](../tutorials/tutorials.md).
+!!! note
+    The `NXmpes` application is an old and trimmed version which does not serve any scientific purpose, but only is used to illustrate the results of the conversion process.
```
```suggestion
    The application definition used here is an old and trimmed version of `NXmpes`  which does not serve any scientific purpose, but is only used to illustrate the results of the conversion process.
```

```suggestion
    The application definition used here is an old and trimmed version of `NXmpes`  which does not serve any scientific purpose, but is only used here to illustrate the results of the conversion process.
```

RubelMozumder replied:
Updated!

lukaspie commented on docs/learn/explanations.md:
```diff
@@ -0,0 +1,8 @@
+# NeXus Definition Language (NXDL), XML, and YAML
+The NeXus data format, described by the NeXus Definition Language (NXDL), represents a concerted effort aimed at facilitating data exchange within scientific communities, particularly among those engaged in neutron, X-ray, and muon research [Könnecke, M. et al.](https://doi.org/10.1107/S1600576714027575). The data format is also being used by the material science community under the project [NeXu ...
```
```suggestion
The NeXus data format, described by the NeXus Definition Language (NXDL), represents a concerted effort aimed at facilitating data exchange within scientific communities, particularly among those engaged in neutron, X-ray, and muon research (see the most recent NeXus paper by [Könnecke et al.](https://doi.org/10.1107/S1600576714027575)). The data format is also used by the material science community under the project [NeXus-FAIRmat proposal](https://fairmat-experimental.github.io/nexus-fairmat-proposal/) supporting FAIR (Findable, Accessible, Interoperable and Reuseable) data principles in material science. The format serves as a standardized framework for both data exchange and storage.
```

RubelMozumder replied:
Addressed!

lukaspie commented on docs/index.md:
```diff
@@ -0,0 +1,52 @@
+An introduction to NeXus and its use in the light of `YAML` format using the `nyaml` tool.
+
+- [An introduction to NeXus](https://manual.nexusformat.org/index.html)
+- [Motivation behind the nyaml tool](learn/explanations.md)
+- [Workflow of the nyaml tool with examples](learn/workflow.md)
```
These must be added to the mkdocs.yaml file in the `nav` section

RubelMozumder replied:
addressed!

lukaspie commented on docs/references/references.md:
```diff
@@ -0,0 +1,15 @@
+2. <a name="NeXus-FAIRmat proposal"></a>[NeXus-FAIRmat proposal](https://fairmat-experimental.github.io/nexus-fairmat-proposal/)
+3. <a name="FAIRmat-NeXus GitHub repository"></a>[FAIRmat-NeXus GitHub repository](https://github.com/FAIRmat-NFDI/nexus_definitions)
+4. <a name= "Official NeXus website"></a>[Official NeXus website](https://manual.nexusformat.org/index.html)
+5. <a name="Könnecke, M. et al."></a>[Könnecke, M. et al.](https://doi.org/10.1107/S1600576714027575)
```
```suggestion
5. <a name="NeXus paper by Könnecke et al."></a>[NeXus paper by Könnecke et al.](https://doi.org/10.1107/S1600576714027575)
```

RubelMozumder replied:
addressed!

lukaspie reviewed (changes requested):
Mostly typos/grammar corrections, but also some things about the documentation build using mkdocs need to be changed

lukaspie reviewed (changes requested):
Mostly typos/grammar corrections, but also some things about the documentation build using mkdocs need to be changed

mkuehbach commented on docs/assets/nomad-plugin-logo.png:
We should just go with the left part of the logo, i.e. the favicon, as nyaml is a standalone activitity rather than a NOMAD plugin. There is no entry point handling in nyaml.

We should also add the NeXus logo here, i.e. either this
https://github.com/nexusformat/NIAC/blob/master/NeXus_Logo/NeXus_Logo.svg or this
https://github.com/nexusformat/NIAC/blob/master/NeXus_Logo/NeXus_Logo_square.svg
Personally leaning to the second one could place it right next to the NOMAD Logo

mkuehbach replied:
Remove then this NOMAD plugin figure

RubelMozumder replied:
Removed!

mkuehbach commented on docs/howtos/howto.md:
```diff
@@ -0,0 +1,45 @@
+# nyaml Installation and Command
```
Getting started with nyaml

RubelMozumder replied:
Addressed!

mkuehbach commented on docs/howtos/howto.md:
```diff
@@ -0,0 +1,45 @@
+`nyaml` is a Python package that is [published on PyPI](https://pypi.org/project/nyaml/).
+
+## How to Install nyaml
+The tool is published to `PyPI` and available for pip install
```
First part of the sentence is redundant suggestion > "To use the tool as is, you should run:"

mkuehbach commented on docs/howtos/howto.md:
```diff
@@ -0,0 +1,45 @@
+# nyaml Installation and Command
+`nyaml` is a Python package that is [published on PyPI](https://pypi.org/project/nyaml/).
+
+## How to Install nyaml
```
I can even see this working without mentioning always nyaml, i.e. "How to install"
"How to use"

RubelMozumder replied:
Addressed!

mkuehbach commented on docs/howtos/howto.md:
````diff
@@ -0,0 +1,45 @@
+$ pre-commit install
+```
+
+## How to Use the nyaml Tool
````
See above "How to use"

RubelMozumder replied:
addressed!

mkuehbach commented on docs/howtos/howto.md:
````diff
@@ -0,0 +1,45 @@
+```
+
+## How to Use the nyaml Tool
+`nyaml` works as a command line tool to convert NeXus application definitions or base classes from `yaml` file format into the `nxdl.xml` file format and vice-versa. The converter can be called by the command
````
from the `nxdl.xml` format to the `yaml` format and vice-versa.

i.e. I suggest to swop the order, typically one starts with a XML file as XML is still the preferred format by the NIAC.

RubelMozumder replied:
addressed!

mkuehbach commented on docs/howtos/howto.md:
````diff
@@ -0,0 +1,45 @@
+                       assist in identifying issues in YAML files.
+  --help               Show this message and exit.
+```
+The `--output-file` option defines the output file name (including the fle extension), otherwise the converter will define the output file name from the input file, e.g., for a input file `NXapplication.nxdl.xml (NXapplication.yaml)`, the resultant file will be `NXapplication_parser.yaml (NXapplication.nxdl.xml)`. With the option `--check-consistency` *the nyaml produces the same type of file as  ...
````
"fle" > "file"
", otherwise" > ". If this option is  not provided, the converter will ..."
"output file name from the input file" > "output file name from the input file name"

", e.g., ... ". Exemplified for an input file named "NXapplication.nxdl.xml", the resultant file will be named ...

Line break before "With the option --check-consistency"

"*the nyaml produces" > "nyaml produces"
for an input file

RubelMozumder replied:
addressed!

mkuehbach commented on docs/howtos/howto.md:
````diff
@@ -0,0 +1,45 @@
+                       assist in identifying issues in YAML files.
+  --help               Show this message and exit.
+```
+The `--output-file` option defines the output file name (including the fle extension), otherwise the converter will define the output file name from the input file, e.g., for a input file `NXapplication.nxdl.xml (NXapplication.yaml)`, the resultant file will be `NXapplication_parser.yaml (NXapplication.nxdl.xml)`. With the option `--check-consistency` *the nyaml produces the same type of file as  ...
````
"nxdl.xml" instead of "nxd.xml"

RubelMozumder replied:
Addressed!

mkuehbach commented on docs/howtos/howto.md:
````diff
@@ -0,0 +1,45 @@
+                       assist in identifying issues in YAML files.
+  --help               Show this message and exit.
+```
+The `--output-file` option defines the output file name (including the fle extension), otherwise the converter will define the output file name from the input file, e.g., for a input file `NXapplication.nxdl.xml (NXapplication.yaml)`, the resultant file will be `NXapplication_parser.yaml (NXapplication.nxdl.xml)`. With the option `--check-consistency` *the nyaml produces the same type of file as  ...
````
"The intention for this option is to verify proper conversion and version conversion of the file." required?

"proper conversion and version conversion" reads a bit weird to me

RubelMozumder replied:
Addressed!

mkuehbach commented on docs/howtos/howto.md:
````diff
@@ -0,0 +1,45 @@
+                       assist in identifying issues in YAML files.
+  --help               Show this message and exit.
+```
+The `--output-file` option defines the output file name (including the fle extension), otherwise the converter will define the output file name from the input file, e.g., for a input file `NXapplication.nxdl.xml (NXapplication.yaml)`, the resultant file will be `NXapplication_parser.yaml (NXapplication.nxdl.xml)`. With the option `--check-consistency` *the nyaml produces the same type of file as  ...
````
The option `--do-not-store-nxdl` prevents the `yaml` file from storing the original `nxdl.xml` text as comment.

We should tell that the default is that the XML part is added in the butt.
Then one can write something like "Setting --do-not-store-nxdl deactivates the addition of the NXDL XML representation to the tail of the YAML file."

RubelMozumder replied:
This are option flags.

mkuehbach commented on docs/index.md:
```diff
@@ -0,0 +1,52 @@
+# FAIRmat nyaml documentation
```
Wouldnt mention FAIRmat here but in the footer where indeed we should acknowledge the funding from the DFG

See an example here:
https://fairmat-nfdi.github.io/pynxtools-em/
bottom of the page with hyperlink to DFG page

mkuehbach replied:
Maybe also a statement supported by the NIAC is useful

RubelMozumder replied:
Addressed!

mkuehbach commented on docs/index.md:
```diff
@@ -0,0 +1,52 @@
+# FAIRmat nyaml documentation
+
+The `nyaml` tool offers a user-friendly command-line interface for converting between `YAML` and `XML` representations of NeXus definitions written in the NeXus Definition Language (NXDL). While the official NeXus definitions are maintained in `XML,` authoring them in `YAML` provides a more concise and readable alternative. `nyaml` bridges this gap, allowing developers to work in `YAML` while ens ...
```
swop order `XML` and `YAML` representations

RubelMozumder replied:
Addressed!

mkuehbach commented on docs/index.md:
```diff
@@ -0,0 +1,52 @@
+
+The `nyaml` tool offers a user-friendly command-line interface for converting between `YAML` and `XML` representations of NeXus definitions written in the NeXus Definition Language (NXDL). While the official NeXus definitions are maintained in `XML,` authoring them in `YAML` provides a more concise and readable alternative. `nyaml` bridges this gap, allowing developers to work in `YAML` while ens ...
+
+The tool supports both application definitions and base classes. It can be installed via `pip` and includes comprehensive documentation and tutorials to assist users in writing NeXus definitions using `YAML`.
```
"The tool supports both application definitions and base classes." why to mention it when it supports all classes? Could left people wondering if contributed_definitions are not supported although they are.

RubelMozumder replied:
improved!

mkuehbach commented on docs/learn/explanations.md:
hyperlink to https://yaml.org/ when writing YAML upon first occurance

mkuehbach replied:
Was this addressed?

RubelMozumder replied:
Addressed!

mkuehbach commented on docs/learn/explanations.md:
```diff
@@ -0,0 +1,8 @@
+# NeXus Definition Language (NXDL), XML, and YAML
+The NeXus data format, described by the NeXus Definition Language (NXDL), represents a concerted effort aimed at facilitating data exchange within scientific communities, particularly among those engaged in neutron, X-ray, and muon research [Könnecke, M. et al.](https://doi.org/10.1107/S1600576714027575). The data format is also being used by the material science community under the project [NeXu ...
```
Supporting the rewriting, we must not point though to outdated links.
Therefore, use [FAIRmat-NFDI ](https://github.com/FAIRmat-NFDI/nexus_definitions) instead of the fairmat-experimental -ish

RubelMozumder replied:
I'm not sure why the `NeXus-FAIRmat` is outdated. The platform where the appdef and base classes are being represented from `nexus-definition` repo.

mkuehbach replied:
Irrelevant why they are outdated, they are, please update the link to as you said the https://fairmat-nfdi.github.io/nexus_definitions/, comment and resolve

mkuehbach commented on docs/learn/explanations.md:
```diff
@@ -0,0 +1,8 @@
+# NeXus Definition Language (NXDL), XML, and YAML
+The NeXus data format, described by the NeXus Definition Language (NXDL), represents a concerted effort aimed at facilitating data exchange within scientific communities, particularly among those engaged in neutron, X-ray, and muon research [Könnecke, M. et al.](https://doi.org/10.1107/S1600576714027575). The data format is also being used by the material science community under the project [NeXu ...
```
"Interoperable, and Reusable"

mkuehbach replied:
Drop "data" in front of "principles"

mkuehbach replied:
Add Wilkinson paper when mentioning FAIR
https://www.nature.com/articles/sdata201618

RubelMozumder replied:
Addressed!

mkuehbach commented on docs/learn/explanations.md:
```diff
@@ -0,0 +1,8 @@
+# NeXus Definition Language (NXDL), XML, and YAML
+The NeXus data format, described by the NeXus Definition Language (NXDL), represents a concerted effort aimed at facilitating data exchange within scientific communities, particularly among those engaged in neutron, X-ray, and muon research [Könnecke, M. et al.](https://doi.org/10.1107/S1600576714027575). The data format is also being used by the material science community under the project [NeXu ...
+
+At the NeXus data format's core, the NeXus Definition Language (NXDL) functions as the cornerstone through which scientists delineate the nomenclature and organizational structure of information within NeXus data files, tailored to specific scientific techniques. NXDL is used to define general data structures (base classes), which define the set of terms that might be used in an instance of that  ...
```
"The NeXus Definition Language (NXDL) at the core of the NeXus data format functions as ..."

mkuehbach replied:
"cornerstone through which" > "whereby"

mkuehbach replied:
pieces of information

mkuehbach replied:
personally would remove ", tailored to ..." cuz NXDL is always used irrespective if class was tailored or not

mkuehbach replied:
remove comma before [XML]

RubelMozumder replied:
addressed!

mkuehbach commented on docs/learn/explanations.md:
```diff
@@ -0,0 +1,8 @@
+
+At the NeXus data format's core, the NeXus Definition Language (NXDL) functions as the cornerstone through which scientists delineate the nomenclature and organizational structure of information within NeXus data files, tailored to specific scientific techniques. NXDL is used to define general data structures (base classes), which define the set of terms that might be used in an instance of that  ...
+
+To expedite the schema development process, we have introduced the use of Yet Another Markup Language ([YAML](https://yaml.org/)), which provides a syntax or style specifically tailored for defining scientific domain-driven schemas with NXDL. One significant advantage of YAML over XML is its indentation-driven approach, which eliminates the need for starting and ending tags for each entity within ...
```
"we have introduced the use of" > "recent the use of ... was introduced"

"The YAML format reduces repetition of NXDL keywords and offers a" > "Thus, using YAML reduces repetition of NXDL keywords and offers a"

RubelMozumder replied:
Addressed!

mkuehbach commented on docs/learn/explanations.md:
```diff
@@ -0,0 +1,8 @@
+
+To expedite the schema development process, we have introduced the use of Yet Another Markup Language ([YAML](https://yaml.org/)), which provides a syntax or style specifically tailored for defining scientific domain-driven schemas with NXDL. One significant advantage of YAML over XML is its indentation-driven approach, which eliminates the need for starting and ending tags for each entity within ...
+
+While `YAML` offers a more readable and concise way to author NeXus definitions, a transcoding mechanism is required because the official format remains XML. The [nyaml](https://github.com/FAIRmat-NFDI/nyaml) Python package serves as a converter tool designed specifically for this purpose. It enables the conversion of NXDL-compliant definitions from YAML to XML, making it easier for NeXus schema  ...
```
"designed specifically for" > "for specifically"

RubelMozumder replied:
Addressed!

mkuehbach commented on docs/references/references.md:
```diff
@@ -0,0 +1,15 @@
+List of references used in this documentation:
+
+1. <a name="nyaml GitHub repository"></a>[nyaml GitHub repository](https://github.com/FAIRmat-NFDI/nyaml)
+2. <a name="NeXus-FAIRmat proposal"></a>[NeXus-FAIRmat proposal](https://fairmat-experimental.github.io/nexus-fairmat-proposal/)
```
We must no longer point to this outdated experimental-ish resource, instead please point to FAIRmat-NFDI/nexus_definitions

RubelMozumder replied:
See previous comment!

mkuehbach replied:
https://fairmat-nfdi.github.io/nexus_definitions/ is to be used

mkuehbach commented on docs/references/references.md:
```diff
@@ -0,0 +1,15 @@
+3. <a name="FAIRmat-NeXus GitHub repository"></a>[FAIRmat-NeXus GitHub repository](https://github.com/FAIRmat-NFDI/nexus_definitions)
+4. <a name= "Official NeXus website"></a>[Official NeXus website](https://manual.nexusformat.org/index.html)
+5. <a name="Könnecke, M. et al."></a>[Könnecke, M. et al.](https://doi.org/10.1107/S1600576714027575)
+6. <a name="XML"></a>[XML](https://www.w3.org/TR/REC-xml/REC-xml-20081126.xml)
```
<a name="XML 1.0 specification"></a>[XML 1.0 specification](https://www.w3.org/TR/REC-xml/REC-xml-20081126.xml)

mkuehbach commented on docs/references/references.md:
```diff
@@ -0,0 +1,15 @@
+6. <a name="XML"></a>[XML](https://www.w3.org/TR/REC-xml/REC-xml-20081126.xml)
+7. <a name="nyaml"></a>[nyaml](https://github.com/FAIRmat-NFDI/nyaml/tree/main)
+8. <a name="NeXus official website"></a>[NeXus official website](https://www.nexusformat.org/)
+9. <a name="YAML"></a>[YAML](https://yaml.org/)
```
<a name="Yet Another Markup Language YAML"></a>[Yet Another Markup Language YAML](https://yaml.org/)

mkuehbach commented on docs/references/references.md:
```diff
@@ -0,0 +1,15 @@
+7. <a name="nyaml"></a>[nyaml](https://github.com/FAIRmat-NFDI/nyaml/tree/main)
+8. <a name="NeXus official website"></a>[NeXus official website](https://www.nexusformat.org/)
+9. <a name="YAML"></a>[YAML](https://yaml.org/)
+10. <a name="lxml"></a>[lxml](https://lxml.de/)
```
10. <a name="lxml - XML and HTML with Python Behnel et al."></a>[lxml - XML and HTML with Python Behnel et al.](https://lxml.de/)

mkuehbach commented on docs/tutorials/tutorials.md:
```diff
@@ -0,0 +1,382 @@
+# Tutorials for NXDL in YAML format
+This tutorial will explain different keywords, terms, and rules from the perspective of YAML format of the NeXus schema. It provides a overall grasp on how to write a NeXus schema (base classes and application definitions) in YAML format using those syntactical components.
+
+## Design of NeXus Ontology and Terms in YAML
```
do not mention Ontology here, the text explains something different

RubelMozumder replied:
Addressed!

mkuehbach commented on docs/tutorials/tutorials.md:
```diff
@@ -0,0 +1,382 @@
+
+## Design of NeXus Ontology and Terms in YAML
+
+Within the YAML format, the root section denotes the top-level description of the application definition or base class schema, comprising the `category`, `type`, `doc`, `symbols` block, and the name of the schema (e.g. `NXmpes(NXobject)`). Correspondingly, the root section refers to the XML element `definition`, encompassing the first `doc` child of the `definition` and `symbols`. The definition  ...
```
"the application definition or the base class schema"

mkuehbach replied:
This covers the `category`, `type`, `doc`, and `symbols` blocks and the name of the schema (e.g. `NXmpes(NXobject)`).

"blocks" or "parameters" unsure about this

mkuehbach replied:
"paranthesis" > "parenthesis"

mkuehbach replied:
make a full-stop before "an application definition may extend"

RubelMozumder replied:
Addressed!

mkuehbach commented on docs/tutorials/tutorials.md:
```diff
@@ -0,0 +1,382 @@
+
+## Design of NeXus Ontology and Terms in YAML
+
+Within the YAML format, the root section denotes the top-level description of the application definition or base class schema, comprising the `category`, `type`, `doc`, `symbols` block, and the name of the schema (e.g. `NXmpes(NXobject)`). Correspondingly, the root section refers to the XML element `definition`, encompassing the first `doc` child of the `definition` and `symbols`. The definition  ...
```
Not immediately but ideal would be to make our punctuation consistent
https://superior-english.com/2017/08/08/comma-after-i-e-and-e-g/

RubelMozumder replied:
Addresed!

mkuehbach commented on docs/tutorials/tutorials.md:
```diff
@@ -0,0 +1,382 @@
+
+## Design of NeXus Ontology and Terms in YAML
+
+Within the YAML format, the root section denotes the top-level description of the application definition or base class schema, comprising the `category`, `type`, `doc`, `symbols` block, and the name of the schema (e.g. `NXmpes(NXobject)`). Correspondingly, the root section refers to the XML element `definition`, encompassing the first `doc` child of the `definition` and `symbols`. The definition  ...
```
"(and XML attribute)" > "(an XML attribute)"

RubelMozumder replied:
Addressed!

mkuehbach commented on docs/tutorials/tutorials.md:
````diff
@@ -0,0 +1,382 @@
+    ```
+### NeXus Group, Field, Attribute, and Link
+#### NeXus Group
+[NeXus groups](https://manual.nexusformat.org/design.html#design-groups), as instances of NeXus base classes, embody the compositional structure of application definitions. Depending on the value of keyword `nameType`, these groups can be initialized dynamically or statically and each approach offers distinct advantages. The keyword `nameType` can hold three district values: `specified`, `any`, a ...
````
as instances of NeXus base classes, and instances of the NXDL/XSD groupType,

RubelMozumder replied:
Addressed!

mkuehbach commented on docs/tutorials/tutorials.md:
````diff
@@ -0,0 +1,382 @@
+    ```
+### NeXus Group, Field, Attribute, and Link
+#### NeXus Group
+[NeXus groups](https://manual.nexusformat.org/design.html#design-groups), as instances of NeXus base classes, embody the compositional structure of application definitions. Depending on the value of keyword `nameType`, these groups can be initialized dynamically or statically and each approach offers distinct advantages. The keyword `nameType` can hold three district values: `specified`, `any`, a ...
````
"can hold three district values: `specified`, `any`, and `partial`" > "can hold one of three distinct values: `specified`, `any`, or `partial`"

RubelMozumder replied:
Addressed!

mkuehbach commented on docs/tutorials/tutorials.md:
````diff
@@ -0,0 +1,382 @@
+
+=== "XML"
+    ```xml
+    <group name="source_TYPE" type="NXsource" recommended="true">
````
add XML attribute nameType="partial"

RubelMozumder replied:
Addressed!

mkuehbach commented on docs/tutorials/tutorials.md:
````diff
@@ -0,0 +1,382 @@
+    </group>
+    ```
+#### NeXus Field and Attribute
+A NeXus `group` may contain NeXus `fields`, NeXus `attributes`, and other NeXus `groups`. A `field`, which is written as a string without a preceding `NX`, and an `attribute`, preceded by `\@`, must have a [NeXus type](https://manual.nexusformat.org/nxdl-types.html#index-0) (e.g.`NX_FLOAT`, `NX_CHAR`). The NeXus type type must be denoted inside parenthesis (e.g. `end_time(NX_DATE_TIME)`); if the  ...
````
A `field` representing an instance of NXDL/XSD fieldType,

mkuehbach replied:
"has an implicit type `NX_CHAR` by default

RubelMozumder replied:
addressed!

mkuehbach commented on docs/tutorials/tutorials.md:
````diff
@@ -0,0 +1,382 @@
+    ```
+
+#### NeXus Link
+The NeXus `link` reduces data duplication since several concepts of the same kind (e.g., NeXus `group`, `field`, or `attribute`) can refer to a single copy of a data element. In YAML format, NeXus `link` is defined denoting the link inside parenthesis. At the same time, the target concept containing the data must be mentioned under the `target` child.
````
"In YAML format, NeXus `link` is defined denoting the link inside parenthesis." > needs editing

RubelMozumder replied:
addressed!

mkuehbach commented on docs/tutorials/tutorials.md:
```diff
@@ -0,0 +1,382 @@
+In the YAML example above, `reference_measurement` is defined as a link refering to an instance of the `NXentry` group with its `target` attribute denoting a value `/entry`. This structure ensures that the concept referencing the data is effectively linked to the designated target, thereby reducing the redundancy and maintaining data integrity within the NeXus framework.
+
+
+<!-- ### NeXus Choice
```
We should at least mention that we do not consistently support this

RubelMozumder replied:
addressed!

RubelMozumder committed:
Implement cspell chcker.

mkuehbach reviewed (changes requested):
See comments

mkuehbach commented on DeleteDownloaded_nyaml_Paper/paper.pdf:
These two files should be removed

mkuehbach commented on DeleteDownloaded_nyaml_Paper/paper.zip:
Remove

mkuehbach commented on docs/learn/workflow.md:
```diff
@@ -0,0 +1,222 @@
+              spec: ISO 18115-1:2023
+              term: 12.58
+              url: https://www.iso.org/obp/ui/en/#iso:std:iso:18115:-1:ed-3:v1:en:term:12.58
+          source_TYPE(NXsource):
```
nameType should be added even though this being an old example

RubelMozumder replied:
Added!

mkuehbach commented on docs/tutorials/tutorials.md:
```diff
@@ -0,0 +1,382 @@
+
+### Special Keywords in YAML
+
+In the YAML schema certain keywords hold self significance beyond their literal interpretations. These special keywords are utilized to elucidate and denote various NeXus terms like `attributes`, `fields`, `links`, and `groups`, thereby enhancing the clarity and specificity of the data representation.
```
Certain keywords hold significance beyond ... to elucidate ...

This improves clarity and specificity

RubelMozumder replied:
Addressed!

mkuehbach commented on docs/tutorials/tutorials.md:
```diff
@@ -0,0 +1,382 @@
+In the YAML schema certain keywords hold self significance beyond their literal interpretations. These special keywords are utilized to elucidate and denote various NeXus terms like `attributes`, `fields`, `links`, and `groups`, thereby enhancing the clarity and specificity of the data representation.
+
+#### Keyword `nameType`
+To initialize a NeXus `group`, `field` or `attribute` the keyword `nameType` carries very significant information on the initialized name depending whether all characters are upper case, lower case or combination of upper-lower case.
```
Invert order "To initialize ... `attribute` and "the keyword `nameType`

i.e. reformulate like this
"The keyword `nameType` carries significant information for defining the name of instances. The following table shows .."

RubelMozumder replied:
reformulated!

mkuehbach commented on docs/tutorials/tutorials.md:
```diff
@@ -0,0 +1,382 @@
+
+#### Keyword `exists`
+
+The `exists` keyword plays a pivotal role in defining the optionality of NeXus concepts `attribute`, `field`, `link`, and `group`. It provides crucial insights for the expected presence or absence of these concepts within the NeXus data structure when implementing the NeXus definitions in NeXus files. By default, all the concepts of a base class are optional, while in an application definition, a ...
```
The keyword `exists`, use parallel structures, i.e. the same way of writing the paragraphs to assist users

mkuehbach replied:
pivotal and crucial using important and only once suffices, see https://www.science.org/doi/10.1126/sciadv.adt3813

RubelMozumder replied:
Yeah, `exists` is a keyword like other keywords and nothing special for attributes, fields, and groups.

mkuehbach replied:
The point was to remove excess words

mkuehbach commented on docs/tutorials/tutorials.md:
````diff
@@ -0,0 +1,382 @@
+      <doc>Voltage applied to detector.</doc>
+    </field>
+    ```
+In the above example, the `detector_voltage` field is defined as a `NX_FLOAT` type with a unit of `NX_VOLTAGE`, indicating that the values stored in this field are measured in volts or millivolts or other voltage units.
````
Do not point that it is NX_FLOAT in this paragraph important is the unit type only

RubelMozumder replied:
I really not see any problem mentioning type `NX_FLOAT`, but the overall paragraph is mainly focused on the `units`.

mkuehbach commented on docs/tutorials/tutorials.md:
```diff
@@ -0,0 +1,382 @@
+
+#### Keyword `dimensions`
+
+The `dimensions` term  describes the multidimensional nature of the data, specifying its rank, dimensional indices, and corresponding length of the rank. The attribute `rank` defines the dimension of the data set. To elucidate each dimension, we use two other keywords: `dim` and `dim_parameters`. The `dim` keyword comprises an array of arrays, the nested array encapsulates values for `index` and  ...
```
rank = "the number of physical dimensions"

mkuehbach replied:
"that each array" 

add "which"  

"corresponds to a keyword"

RubelMozumder replied:
Improved!

mkuehbach commented on docs/tutorials/tutorials.md:
````diff
@@ -0,0 +1,382 @@
+    ```
+
+#### Keyword `enumeration`
+List of strings which are considered as a set of predefined values for fields or attributes. Individual items of the enumeration may also hold a `doc` keyword to provide a description of the item.
````
for `field` and attribute` hyphenation like in the other paragraphs that explain keywords for the sake of consistency

RubelMozumder replied:
`enumeration` is a keyword like other keywords; nothing differs for NeXus `field` `attribute`. I think it's fine current explanations.

mkuehbach commented on docs/tutorials/tutorials.md:
````diff
@@ -0,0 +1,382 @@
+    </field>
+    ```
+
+`open_enum` says along with the listed items other items are also valid while initializing the NeXus object.
````
The parameter `open_enum` defines that also other items are valid

RubelMozumder replied:
Addressed!

mkuehbach reviewed (changes requested):
completes feedback from @mkuehbach

RubelMozumder committed:
adding logo for nyaml.

RubelMozumder committed:
Split navigation sidebar.

RubelMozumder committed:
nyaml workflow diagram.

RubelMozumder committed:
PR comments.

RubelMozumder committed:
Update logo images.

RubelMozumder committed:
remove unwanted files.

RubelMozumder committed:
PR comments.

Resolve PR comments.

RubelMozumder committed:
Deploy the docs from MkDoc branch.

RubelMozumder committed:
Update the branch name.

RubelMozumder committed:
rebase some test commits

RubelMozumder committed:
cspell

ci

test with spell github action image.

RubelMozumder committed:
Resolve the PR comments.

lukaspie commented on docs/howtos/howto.md:
````diff
@@ -0,0 +1,51 @@
+    :style: table
+    :list_subcommands: True
+```
+with the available options:
````
We don't need this anymore because we get it automatically from mkdocs-click. Please remove.

mkuehbach replied:
Trivial thing, was it removed, please feedback and resolve, thank you

RubelMozumder replied:
modified. Yeah, mkdocs-click brings the list of options with command line descriptions. But these are not sufficient added small interpretation fot those options.

lukaspie commented on docs/howtos/howto.md:
````diff
@@ -0,0 +1,51 @@
+                       assist in identifying issues in YAML files.
+  --help               Show this message and exit.
+```
+`--output-file`: The option defines the output file name (including the file extension), if the option is not specified the converter will define the output file name from the input file. Exemplified for a given input file `NXapplication.nxdl.xml (NXapplication.yaml)`, the resultant file will be `NXapplication_parser.yaml (NXapplication.nxdl.xml)`.
````
Why not make this a list?

mkuehbach replied:
Leaving the ```--help`` environment as is okay as well.
Followed by one blank line and one paragraph for each option, i.e.
Line 45 indent by one line

RubelMozumder replied:
included!

lukaspie commented on docs/learn/explanations.md:
```diff
@@ -0,0 +1,8 @@
+
+The NeXus Definition Language (NXDL) at the core of the NeXus data format functions as the cornerstone whereby scientists delineate the nomenclature and organizational pieces of information within NeXus data files specific to the scientific techniques. NXDL is used to define general data structures (base classes), which define the set of terms that might be used in an instance of that class. Thes ...
+
+To expedite the schema development process, recently, we have introduced the use of Yet Another Markup Language ([YAML](https://yaml.org/)), which provides a syntax or style specifically tailored for defining scientific domain-driven schemas with NXDL. One significant advantage of YAML over XML is its indentation-driven approach, which eliminates the need for starting and ending tags for each ent ...
```
It isn't called Yet Another Markup Language anymore, maybe just use the acronym
```suggestion
To expedite the schema development process, recently, we have introduced the use of [YAML](https://yaml.org/), which provides a syntax or style specifically tailored for defining scientific domain-driven schemas with NXDL. One significant advantage of YAML over XML is its indentation-driven approach, which eliminates the need for starting and ending tags for each entity within the schema. Thus, the use of YAML format reduces repetition of NXDL keywords and offers a more intuitive structure aligned with object-oriented programming concepts, such as class inheritance. These benefits are attained without compromising the integrity of the original NeXus schema, which is traditionally expressed in XML format.
```

mkuehbach replied:
Trivial thing, fixed? if so, mention and resolve

lukaspie commented on docs/learn/explanations.md:
```diff
@@ -0,0 +1,8 @@
+
+To expedite the schema development process, recently, we have introduced the use of Yet Another Markup Language ([YAML](https://yaml.org/)), which provides a syntax or style specifically tailored for defining scientific domain-driven schemas with NXDL. One significant advantage of YAML over XML is its indentation-driven approach, which eliminates the need for starting and ending tags for each ent ...
+
+While `YAML` offers a more readable and concise way to author NeXus definitions, a transcoding mechanism is required because the official format remains XML. The [nyaml](https://github.com/FAIRmat-NFDI/nyaml) Python package serves as a converter tool designed for specifically this purpose. It enables the conversion of NXDL-compliant definitions from YAML to XML, making it easier for NeXus schema  ...
```
Decide if you want to write plain "YAML" or "`YAML`"

mkuehbach replied:
This is a point about consistent writing, trivial fix:
1.) Make a choice e.g. `YAML`
2.) Search for occurrences and replace by above-choice
3.) Leave a comment like fixed and done

RubelMozumder replied:
Kept `YAML` and fixed the inconsistency over the md files.

RubelMozumder committed:
fix mkdocs-click and options.

RubelMozumder committed:
PR comment resolve.

PR comment resolve.

RubelMozumder committed:
update link

RubelMozumder commented on .github/workflows/build_docs.yml:
```diff
@@ -0,0 +1,34 @@
+name: build_and_deploy_docs
+on:
+  push:
+    branches: [MyDoc]
```
Revert it to `main`

lukaspie reviewed (approved):
lgtm

mkuehbach reviewed (approved):
a reasonable set of changes

RubelMozumder committed:
Upgrade to main branch.

RubelMozumder committed:
Update workflow.

Minor fix and remove irrelevent files.
